history-network.md

```diff
@@ -0,0 +1,300 @@
+
+The history protocol uses the node table structure from the Discovery v5 network and the lookup algorithm from section 2.3 of the Kademlia paper.
+
+### Data
+
+Types:
+
+* Block headers
+* Block bodies
+    * Transactions
+    * Receipts
+
+Lookups:
+
+* Block header by block header hash
+* Block body by block header hash
+* Block receipts by block header hash
+
+This specification does not support block lookups by number or transaction lookups by hash. To support these lookups, we will require a specification for a block header accumulator so that we can return the canonical block for a given height or the transaction for a given hash included in some canonical block.
+
+### Data Bridge
+
+Participants in the `eth` protocol are responsible for injecting data into the network. These nodes act as a bridge between the `eth` protocol and the chain history network. Once the data is present within the network, that data is distributed to nodes "close" to that data. We define a notion of distance between some node and some data item below.
+
+### Data Completeness
+In order to ensure that the network holds the entirety of the chain history, a separate process searches for missing data. Upon discovery of some missing data, that node issues a request to a bridge node for that data. The bridge node responds with the data, and it is distributed in the same way that new history is distributed.
+
+## Specification
+
+### Distance
+
+Nodes in the history network are represented by their [EIP-778 Ethereum Node Record (ENR)](https://eips.ethereum.org/EIPS/eip-778) from the Discovery v5 network. A node's `node-id` is derived according to the node's identity scheme, which is specified in the node's ENR. A node's `node-id` represents its address in the DHT.
+
+The `node-id` is a 32-byte identifier. We define the `distance` function that maps a pair of `node-id` values to a 256-bit unsigned integer identically to the Discovery v5 network.
+
+```
+distance(n1, n2) = n1 XOR n2
+```
+
+Similarly, we define a `logdistance` function identically to the Discovery v5 network.
+
+```
+logdistance(n1, n2) = log2(distance(n1, n2))
+```
+
+### Content: Keys and Values
+
+The chain history DHT stores the following data items:
+
+* Block headers
+* Block bodies
+* Receipts
+
+Each of these data items are represented as a key-value pair. Denote the key for a data item by `content-key`. Denote the value for an item as `content`.
+
+We derive a `content-id` from the `content-key` as `H(content-key)` where `H` denotes the [placeholder] hash function, which outputs 32-byte values. The `content-id` represents the key in the DHT that we use for `distance` calculations.
+
+All `content-key` values are encoded and decoded according to SSZ sedes.
+
+#### Block Header
+
+```
+content-key  = Container(chain-id: uint16, content-type: uint8, block-hash: Bytes32)
```

Comment:
Fun, I almost left a comment suggesting we just define this as a custom encoding using concatenation rather than bringing SSZ into this, but I realize the two are equivalent since these are fixed size types and SSZ encoding of this type is in effect simply concatenation.